src/engine/channels/enginedeck.cpp

```diff
@@ -39,39 +46,110 @@ EngineDeck::EngineDeck(
+                ConfigKey(getGroup(), QString("stem_%1_volume").arg(stemIdx))));
+        // The default value is ignored and override with the medium value by
+        // ControlPotmeter. This is likely a bug but fixing might have a
+        // disruptive impact, so setting the default explicitly
+        m_stemGain.back()->set(1.0);
+        m_stemGain.back()->setDefaultValue(1.0);
+        m_stemMute.emplace_back(std::make_unique<ControlPushButton>(
+                ConfigKey(getGroup(), QString("stem_%1_mute").arg(stemIdx))));
+    }
+#endif
 }
 
+#ifdef __STEM__
+void EngineDeck::slotTrackLoaded(TrackPointer pNewTrack,
+        TrackPointer) {
+    VERIFY_OR_DEBUG_ASSERT(m_pStemCount) {
+        return;
+    }
+    if (m_pConfig->getValue(
+                ConfigKey("[Mixer Profile]", "stem_auto_reset"), true)) {
+        for (int stemIdx = 0; stemIdx < kMaxSupportedStem; stemIdx++) {
+            m_stemGain[stemIdx]->set(1.0);
+            m_stemMute[stemIdx]->set(0.0);
+            ;
+        }
+    }
+    if (pNewTrack) {
+        int stemCount = pNewTrack->getStemInfo().size();
+        m_pStemCount->forceSet(stemCount);
+    } else {
+        m_pStemCount->forceSet(0);
+    }
+}
+#endif
+
 EngineDeck::~EngineDeck() {
     delete m_pPassing;
     delete m_pBuffer;
     delete m_pPregain;
 }
 
+#ifdef __STEM__
 void EngineDeck::processStem(CSAMPLE* pOut, const int iBufferSize) {
-    int stereoChannelCount = m_pBuffer->getChannelCount() / mixxx::kEngineChannelOutputCount;
-    auto allChannelBufferSize = iBufferSize * stereoChannelCount;
+    int stemCount = m_pBuffer->getChannelCount() / mixxx::kEngineChannelOutputCount;
+    auto allChannelBufferSize = iBufferSize * stemCount;
     if (m_stemBuffer.size() < allChannelBufferSize) {
         m_stemBuffer = mixxx::SampleBuffer(allChannelBufferSize);
     }
     m_pBuffer->process(m_stemBuffer.data(), allChannelBufferSize);
 
     // TODO(XXX): process effects per stems
-
     SampleUtil::clear(pOut, iBufferSize);
     const CSAMPLE* pIn = m_stemBuffer.data();
     for (int i = 0; i < iBufferSize; i += mixxx::kEngineChannelOutputCount) {
-        for (int chIdx = 0; chIdx < m_pBuffer->getChannelCount();
-                chIdx += mixxx::kEngineChannelOutputCount) {
-            // TODO(XXX): apply stem gain or skip muted stem
-            pOut[i] += pIn[stereoChannelCount * i + chIdx];
-            pOut[i + 1] += pIn[stereoChannelCount * i + chIdx + 1];
+        for (int stemIdx = 0; stemIdx < stemCount;
+                stemIdx++) {
+            if (m_stemMute[stemIdx]->toBool()) {
+                continue;
+            }
+            float gain = static_cast<float>(m_stemGain[stemIdx]->get());
+            pOut[i] += pIn[stemCount * i + mixxx::kEngineChannelOutputCount * stemIdx] * gain;
+            pOut[i + 1] +=
+                    pIn[stemCount * i +
+                            mixxx::kEngineChannelOutputCount * stemIdx + 1] *
+                    gain;
```

Comment:
I expect suggestion here about factoring this code to help vectorization and move it into `SampleUtil`. Please note this is already done as part of the [next iteration](https://github.com/acolombier/mixxx/blob/feat/add-quick-fx-on-stem/src/engine/channels/enginedeck.cpp#L144)